hmmlearn/hmm.py

```diff
@@ -520,134 +525,294 @@ class GMMHMM(_BaseHMM):
+                                         .format(j, i))
 
     def _generate_sample_from_state(self, state, random_state=None):
-        return self.gmms_[state].sample(1, random_state=random_state).flatten()
+        # TODO: actually use random_state
+        if random_state is None:
+            random_state = self.random_state
+        random_state = check_random_state(random_state)
+
+        cur_means = self.means_[state]
+        cur_covs = self.covars_[state]
+        cur_weights = self.weights_[state]
+
+        i_gauss = random_state.choice(self.n_mix, p=cur_weights)
+        mean = cur_means[i_gauss]
+        if self.covariance_type == 'tied':
+            cov = cur_covs
+        else:
+            cov = cur_covs[i_gauss]
+
+        return sample_gaussian(mean, cov, self.covariance_type,
+                               random_state=random_state)
+
+    def _compute_log_weighted_gaussian_densities(self, X, i_comp):
+        cur_means = self.means_[i_comp]
+        cur_covs = self.covars_[i_comp]
+        if self.covariance_type == 'spherical':
+            cur_covs = cur_covs[:, np.newaxis]
+        log_cur_weights = np.log(self.weights_[i_comp])
+
+        return log_multivariate_normal_density(
+            X, cur_means, cur_covs, self.covariance_type
+        ) + log_cur_weights
+
+    def _compute_log_likelihood(self, X):
+        n_samples, _ = X.shape
+        res = np.zeros((n_samples, self.n_components))
+
+        for i in range(self.n_components):
+            log_denses = self._compute_log_weighted_gaussian_densities(X, i)
+            res[:, i] = logsumexp(log_denses, axis=1)
+
+        return res
 
     def _initialize_sufficient_statistics(self):
         stats = super(GMMHMM, self)._initialize_sufficient_statistics()
-        stats['norm'] = [np.zeros(g.weights_.shape) for g in self.gmms_]
-        stats['means'] = [np.zeros(np.shape(g.means_)) for g in self.gmms_]
-        stats['covars'] = [np.zeros(np.shape(g.covars_)) for g in self.gmms_]
+        stats['n_samples'] = 0
+        stats['post_comp_mix'] = None
+        stats['post_mix_sum'] = np.zeros((self.n_components, self.n_mix))
+        stats['post_sum'] = np.zeros(self.n_components)
+        stats['samples'] = None
+        stats['centered'] = None
         return stats
 
     def _accumulate_sufficient_statistics(self, stats, X, framelogprob,
-                                          posteriors, fwdlattice, bwdlattice):
+                                          post_comp, fwdlattice, bwdlattice):
+
+        # TODO: support multiple frames
+
         super(GMMHMM, self)._accumulate_sufficient_statistics(
-            stats, X, framelogprob, posteriors, fwdlattice, bwdlattice)
+            stats, X, framelogprob, post_comp, fwdlattice, bwdlattice
+        )
+
+        n_samples, _ = X.shape
+
+        stats['n_samples'] = n_samples
+        stats['samples'] = X
+
+        # post_comp = posteriors.reshape((n_samples,))
+
+        eps = 1e-15
```

Comment:
You could use [machine epsilon](http://stackoverflow.com/a/19141711/262432) here.